**Postgrado en Ciencias de la Computación**  
**Inteligencia Artificial (CC6021236)**  
**Semestre 2019-1**  
**Prof. Eugenio Scalise**  

## Tarea 1 (Técnicas de búsqueda Heurística)

El 8-puzzle es un juego que consiste en un tablero de 3x3 casillas, donde una de las casillas está vacía y el resto están ocupadas y numeradas desde 1 hasta 8 como se muestra en la siguiente figura:

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | &nbsp; | 3 |
| 5 | 2 | 6 |
| 4 | 7 | 8 |

El propósito es mover las casillas hasta llegar a un estado objetivo. En nuestro caso se usará como objetivo el estado mostrado en la siguiente figura:

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | 2 | 3 |
| 4 | 5 | 6 |
| 7 | 8 | &nbsp; |

Para un estado o posición cualquiera del 8-puzzle, los únicos movimientos válidos son cambiar la casilla vacía con cualquiera de sus adyacentes horizontales o verticales. 

Se desea que realice las siguientes actividades:


1. Proponga una representación para los estados del 8-puzzle, indique el tipo y ejemplos en Python.

Para la representación de los estados utilicé una clase llamada Node, en donde, cada instancia representa un nodo del arbol y un estado del 8-puzzle. Esta clase Node tiene los siguientes atributos:


*   **value:** contiene una matriz de enteros con el estado actual del tablero, por ejemplo: [[1,9,3],[5,2,6],[4,7,8]].
*   **pos:** contiene un lista de enteros de solo dos elementos, el primer elemento es la posición en filas del "blanco" (para efectos del código se manejó como el número 9) y el segundo elemento es la posición en columnas del "blanco".
*   **h:** contiene el valor de f(n) = g(n) + h(n) de cada estado del 8-puzzle.
*   **level:** contiene el valor entero que representa el nivel del nodo en el árbol.
*   **parent:** contiene un apuntador al nodo padre.
*   **child:** contiene un arreglo de apuntadores a los nodos hijos.

**Ejemplos:** para los siguientes ejemplos de estados del 8-puzzle se muestra cómo queda instaciado el nodo de la clase Node en cada caso:

**1)**

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | &nbsp; | 3 |
| 5 | 2 | 6 |
| 4 | 7 | 8 |


root = Node([[1,9,3],[5,2,6],[4,7,8]], [0,1], 0, 0, None, [node1,node2,node3])

**2)**

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | 2 | 3 |
| 5 | &nbsp; | 6 |
| 4 | 7 | 8 |


node1 = Node([[1,2,3],[5,9,6],[4,7,8]], [1,1], 1 + 4, 1, root, None)

**3)**

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | 3 | &nbsp; |
| 5 | 2 | 6 |
| 4 | 7 | 8 |

node2 = Node([[1,3,9],[5,2,6],[4,7,8]], [0,2], 1 + 6, 1, root, None)

**4)**

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| &nbsp; | 1 | 3 |
| 5 | 2 | 6 |
| 4 | 7 | 8 |

node3 = Node([[9,1,3],[5,2,6],[4,7,8]], [0,0], 1 + 6, 1, root, None)




2. Defina una función de costo y una heurística para su representación. ¿Es su heurística admisible? (Justifique su respuesta)

El costo de un estado a otro es representado como el costo de realizar el movimiento en el 8-puzzle, es decir, 1 movimiento por lo que el valor del costo de un estado al otro es de 1. Para obtener el costo de un estado después de **n** movimientos simplemente se toma el nivel del padre (el nivel representa la cantidad de movimientos realizados hasta ese estado) y se le suma 1, ya que sería el nivel de ese estado actual o hijo que representa la cantidad de movimientos a realizar para llegar a ese estado.

Para la función heurística se utilizó la distancia Manhattan desde cada posición actual hasta su posición correcta [2], por ejemplo:

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | &nbsp; | 3 |
| 5 | 2 | 6 |
| 4 | 7 | 8 |

Para el caso mostrado la heurística sería:

*   **Para 1:** |1 - 1| + |1 - 1| = 0 + 0 = 0
*   **Para 2:** |1 - 2| + |2 - 2| = 1 + 0 = 1
*   **Para 3:** |1 - 1| + |3 - 3| = 0 + 0 = 0
*   **Para 4:** |2 - 3| + |1 - 1| = 1 + 0 = 1
*   **Para 5:** |2 - 2| + |2 - 1| = 0 + 1 = 1
*   **Para 6:** |2 - 2| + |3 - 3| = 0 + 0 = 0
*   **Para 7:** |3 - 3| + |1 - 2| = 0 + 1 = 1
*   **Para 8:** |3 - 3| + |2 - 3| = 0 + 1 = 1

Cuando se realiza la suma de cada una se tiene que: 

h(n) = 0 + 1 + 0 + 1 + 1 + 0 + 1 + 1 = 5

Tomando en cuenta que [2]:

*  n es un nodo.
*  h es una heurística.
*  h(n) es el la estimación del coste mínimo para alcanzar el objetivo desde n, aplicando h.
*  C(n) es el coste mínimo real para alcanzar el objetivo desde n.

Decimos que h es admisible si para todo n, h(n) <= C(n).

Determinamos que la heurística usada es admisible ya que: C(n) es el coste mínimo para alcanzar desde n al objetivo y la distancia Manhattan calcula la distancia mínima que se puede realizar de un estado al otro, con lo cual siempre se cumplirá para cualquier n en cualquier posición que: h(n) <= C(n). Por ejemplo: para el ejemplo anterior y para el número 2 tenemos que: h(2) = 1 y C(2) = 1, por lo tanto h(2) <= C(2) se cumple.

Escogí esta heurística debido a que busca siempre obtener las distancias de cada número desde su posición actual a su posición original o correcta, por lo que al evaluar f(n) = g(n) + h(n) se procupara escoger el estado en donde sus números estan más cerca a sus posiciones correctas.

In [0]:
def cost(parent_cost):
  return parent_cost + 1

def heuristic(matrix):
     total = 0
     h = 0
     
     for i in range(n):
          for j in range(n):
               if matrix[i][j] == 1:
                    h = abs(0 - i) + abs(0 - j)
               if matrix[i][j] == 2:
                    h = abs(0 - i) + abs(1 - j)
               if matrix[i][j] == 3:
                    h = abs(0 - i) + abs(2 - j)
               if matrix[i][j] == 4:
                    h = abs(1 - i) + abs(0 - j)
               if matrix[i][j] == 5:
                    h = abs(1 - i) + abs(1 - j)
               if matrix[i][j] == 6:
                    h = abs(1 - i) + abs(2 - j)
               if matrix[i][j] == 7:
                    h = abs(2 - i) + abs(0 - j)
               if matrix[i][j] == 8:
                    h = abs(2 - i) + abs(1 - j)
               if matrix[i][j] == 9:
                    h = 0
               total = total + h
     
     return total

3. Implemente una solución al problema 8-puzzle utilizando el algoritmo A*, considerando la representación y heurística definidas anteriormente.

Para efectos de análisis en próximas preguntas es importante que incorpore a su solución un conteo del número de estados visitados durante la búsqueda, el promedio de anchura/hijos del árbol de búsqueda, el tamaño de la solución obtenida y el tiempo de ejecución (para esto último utilice el *comando mágico* %%time de Jupyter).

Es importante destacar que en el código y toda implementación manejo el "blanco" en el 8-puzzle como el número 9. 

Además, otro punto importante es que antes de llamar o ejecutar las funciones necesarias para obtener la solución a través del algoritmo de A*, primero se ejecuta una función llamada **validateMatrix()**, la cual permite verificar si el tablero ingresado es un tablero válido o si por el contrario no tiene solución. En caso de que sea válido y tenga solución, ejecuta el algoritmo para resolver el problema.

Las ejecuciones del programa se realiza llamando a la función **mainA()** y se le pasa como parametro la matriz con el tablero del caso de prueba. Por ejemplo: mainA([[2,3,8],[7,4,1],[5,6,9]])

In [0]:
repeat = 0
n = 0
depth_max = 0
front_count = 0

class Node:
     def __init__(self,value,pos,h,level,parent,child):
          self.value = value
          self.pos = pos
          self.h = h
          self.level = level
          self.parent = parent
          self.child = child

def find(child,front,exp):
     for i in front:
          if i.value == child.value:
               return "f"
     for j in exp:
          if j.value == child.value:
               return "e"
          
     return "n"

def replaceFront(old,child):
     new = []
     inserted = False

     for i in old:
          if i.value == child.value:
               if child.h < i.h and not(inserted):
                    new.append(child)
          elif child.h < i.h and not(inserted):
               new.append(child)
               new.append(i)
               inserted = True
          else:
               new.append(i)
               
     return new

def replaceExp(old,child):
     new = []

     for i in old:
          if i.value == child.value and child.h < i.h:
               new.append(child)
          else:
               new.append(i)
               
     return new
    
def insert(old,child):
     new = []
     inserted = False
     x = 0
     
     if not(old):
          new.append(child)
          inserted = True
     else:
          for i in old:
               if child.h < i.h and not(inserted):
                    new.append(child)
                    inserted = True
               new.append(i)
               x = x + 1

     if not(inserted):
          new.append(child)
          
     return new

def showMatrix(matrix):
     concat = ""
     
     for i in range(n):
          for j in range(n):
               if matrix[i][j] == 9:
                    concat =  concat + "  "
               else:
                    concat = concat + str(matrix[i][j]) + " "
          concat = concat + "\n"

     print(concat)

def insertMatrix(matrix):
     new = createMatrix()
     
     for i in range(n):
     	for j in range(n):
     		new[i][j] = matrix[i][j]
     
     return new

def createMatrix():
     matrix = [None] * n
     for i in range(n):
         matrix[i] = [None] * n
     return matrix

def createNodes(matrix):
     value = matrix.value
     level = matrix.level     
     i = matrix.pos[0]
     j = matrix.pos[1]
      
     global depth_max
    
     if depth_max < level + 1:
        depth_max = level + 1

     nodes = []

     node = None
     node1 = None
     node2 = None
     node3 = None

     if i == 0:
          if matrix.parent == None:
               node = Node(createMatrix,[i+1,j],0,cost(level),matrix,[])
               node.value = insertMatrix(value)
               node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
               node.value[node.pos[0]][node.pos[1]] = 9
               node.h = heuristic(node.value) + cost(level)
               nodes.append(node)
               
          elif matrix.parent.pos[0] != (i+1):
               node = Node(createMatrix,[i+1,j],0,cost(level),matrix,[])
               node.value = insertMatrix(value)
               node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
               node.value[node.pos[0]][node.pos[1]] = 9
               node.h = heuristic(node.value) + cost(level)
               nodes.append(node)
          
     if i == n - 1:
          if matrix.parent == None:
               node = Node(createMatrix,[i-1,j],0,cost(level),matrix,[])
               node.value = insertMatrix(value)
               node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
               node.value[node.pos[0]][node.pos[1]] = 9
               node.h = heuristic(node.value) + cost(level)
               nodes.append(node)

          elif matrix.parent.pos[0] != (i-1):
               node = Node(createMatrix,[i-1,j],0,cost(level),matrix,[])
               node.value = insertMatrix(value)
               node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
               node.value[node.pos[0]][node.pos[1]] = 9
               node.h = heuristic(node.value) + cost(level)
               nodes.append(node)

     if i > 0 and i < n - 1:
          if matrix.parent != None:
               if matrix.parent.pos[0] != (i+1):
                    node = Node(createMatrix,[i+1,j],0,cost(level),matrix,[])
                    node.value = insertMatrix(value)
                    node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
                    node.value[node.pos[0]][node.pos[1]] = 9
                    node.h = heuristic(node.value) + cost(level)
                    nodes.append(node)

               if matrix.parent.pos[0] != (i-1):
                    node1 = Node(createMatrix,[i-1,j],0,cost(level),matrix,[])
                    node1.value = insertMatrix(value)
                    node1.value[i][j] = node1.value[node1.pos[0]][node1.pos[1]]
                    node1.value[node1.pos[0]][node1.pos[1]] = 9
                    node1.h = heuristic(node1.value) + cost(level)
                    nodes.append(node1)
          else:
               node = Node(createMatrix,[i+1,j],0,cost(level),matrix,[])
               node.value = insertMatrix(value)
               node.value[i][j] = node.value[node.pos[0]][node.pos[1]]
               node.value[node.pos[0]][node.pos[1]] = 9
               node.h = heuristic(node.value) + cost(level)
               nodes.append(node)

               node1 = Node(createMatrix,[i-1,j],0,cost(level),matrix,[])
               node1.value = insertMatrix(value)
               node1.value[i][j] = node1.value[node1.pos[0]][node1.pos[1]]
               node1.value[node1.pos[0]][node1.pos[1]] = 9
               node1.h = heuristic(node1.value) + cost(level)
               nodes.append(node1)

     if j == 0:
          if matrix.parent == None:
               node2 = Node(createMatrix,[i,j+1],0,cost(level),matrix,[])
               node2.value = insertMatrix(value)
               node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
               node2.value[node2.pos[0]][node2.pos[1]] = 9
               node2.h = heuristic(node2.value) + cost(level)
               nodes.append(node2)
               
          elif matrix.parent.pos[1] != (j+1):
               node2 = Node(createMatrix,[i,j+1],0,cost(level),matrix,[])
               node2.value = insertMatrix(value)
               node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
               node2.value[node2.pos[0]][node2.pos[1]] = 9
               node2.h = heuristic(node2.value) + cost(level)
               nodes.append(node2)

     if j == n - 1:
          if matrix.parent == None:
               node2 = Node(createMatrix,[i,j-1],0,cost(level),matrix,[])
               node2.value = insertMatrix(value)
               node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
               node2.value[node2.pos[0]][node2.pos[1]] = 9
               node2.h = heuristic(node2.value) + cost(level)
               nodes.append(node2)

          elif matrix.parent.pos[1] != (j-1):
               node2 = Node(createMatrix,[i,j-1],0,cost(level),matrix,[])
               node2.value = insertMatrix(value)
               node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
               node2.value[node2.pos[0]][node2.pos[1]] = 9
               node2.h = heuristic(node2.value) + cost(level)
               nodes.append(node2)

     if j > 0 and j < n - 1:
          if matrix.parent != None:
               if matrix.parent.pos[1] != (j+1):
                    node2 = Node(createMatrix,[i,j+1],0,cost(level),matrix,[])
                    node2.value = insertMatrix(value)
                    node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
                    node2.value[node2.pos[0]][node2.pos[1]] = 9
                    node2.h = heuristic(node2.value) + cost(level)
                    nodes.append(node2)

               if matrix.parent.pos[1] != (j-1):
                    node3 = Node(createMatrix,[i,j-1],0,cost(level),matrix,[])
                    node3.value = insertMatrix(value)
                    node3.value[i][j] = node3.value[node3.pos[0]][node3.pos[1]]
                    node3.value[node3.pos[0]][node3.pos[1]] = 9
                    node3.h = heuristic(node3.value) + cost(level)
                    nodes.append(node3)
          else:
               node2 = Node(createMatrix,[i,j+1],0,cost(level),matrix,[])
               node2.value = insertMatrix(value)
               node2.value[i][j] = node2.value[node2.pos[0]][node2.pos[1]]
               node2.value[node2.pos[0]][node2.pos[1]] = 9
               node2.h = heuristic(node2.value) + cost(level)
               nodes.append(node2)

               node3 = Node(createMatrix,[i,j-1],0,cost(level),matrix,[])
               node3.value = insertMatrix(value)
               node3.value[i][j] = node3.value[node3.pos[0]][node3.pos[1]]
               node3.value[node3.pos[0]][node3.pos[1]] = 9
               node3.h = heuristic(node3.value) + cost(level)
               nodes.append(node3)

     return nodes

def goal(matrix):
     if matrix != None:
          x = 1

          for i in range(n):
               for j in range(n):
                    if matrix[i][j] != x:
                         return False
                    x = x + 1

          return True

     else:
          return False
      
def validateMatrix(matrix,pos):
     result = 0
     distance = 0

     for i in range(n):
          for j in range(n):
               result = result + count(matrix,matrix[i][j],[i,j])

     distance = abs(n - 1 - pos[0]) + abs(n - 1 - pos[1])

     result = result + distance

     return result % 2 == 0

def count(matrix,value,pos):
     result = 0

     for x in range(n):
          if x >= pos[1] and value > matrix[pos[0]][x]:
               result = result + 1

     for i in range(n):
          for j in range(n):
               if i > pos[0] and value > matrix[i][j]:
                    result = result + 1
                    
     return result

In [0]:
def A_star(matrix,pos):
     node = Node(matrix,pos,0,0,None,[])
     front = [node]
     exp = []
     finded = ""
     global repeat
     global front_count

     while front:
          repeat = repeat + 1
          node = front.pop(0)
    
          if goal(node.value):
               front_count = len(front)
               return node
          
          node.child = createNodes(node)
          exp.append(node)

          for i in node.child:
               finded = find(i,front,exp)
               
               if finded == "n":
                    front = insert(front,i)
               elif finded == "f":
                    front = replaceFront(front,i)
               else:
                    exp = replaceExp(exp,i)


In [68]:
def mainA(matrix):

     global n
     global repeat
     global depth_max
     global front_count
      
     repeat = 0
     n = 3
     depth_sol = 0
     depth_max = 0
     front_count = 0

     pos = []

     for i in range(n):
          for j in range(n):
               if matrix[i][j] == 9:
                    pos = [i,j]
                    
     if validateMatrix(matrix,pos):
          node = A_star(matrix,pos)

          if(node != None):
               result = []
              
               depth_sol = node.level

               while node != None:
                    result.append(node)
                    node = node.parent

               result.reverse()

               print("Camino óptimo:")
               for i in result:
                    showMatrix(i.value)

               print("Movimientos:",len(result) - 1)
               print("Estados explorados:",repeat)
               print("Profundidad de la solución:",depth_sol)
               print("Profundidad máxima:",depth_max)
               print("Tamaño lista de nodos frontera:",front_count)
          else:
               print("No se encontró solución")
     else:
          print("Tablero invalido")
        
#CASOS DE PRUEBA:
print("Caso 1:")
%time mainA([[2,3,8],[7,4,1],[5,6,9]])
print("Caso 2:")
%time mainA([[2,3,7],[1,4,8],[5,6,9]])
print("Caso 3:")
%time mainA([[9,7,6],[1,2,8],[5,3,4]])
print("Caso 4:")
%time mainA([[3,8,6],[9,2,4],[7,1,5]])
print("Caso 5:")
%time mainA([[1,9,3],[5,2,6],[4,7,8]])
print("Caso 6:")
%time mainA([[7,2,4],[5,9,6],[8,3,1]])
print("Caso 7:")
%time mainA([[1,2,3],[8,9,4],[7,6,5]])


Caso 1:
Camino óptimo:
2 3 8 
7 4 1 
5 6   

2 3 8 
7 4   
5 6 1 

2 3   
7 4 8 
5 6 1 

2   3 
7 4 8 
5 6 1 

2 4 3 
7   8 
5 6 1 

2 4 3 
7 6 8 
5   1 

2 4 3 
7 6 8 
5 1   

2 4 3 
7 6   
5 1 8 

2 4 3 
7   6 
5 1 8 

2 4 3 
7 1 6 
5   8 

2 4 3 
7 1 6 
  5 8 

2 4 3 
  1 6 
7 5 8 

2 4 3 
1   6 
7 5 8 

2   3 
1 4 6 
7 5 8 

  2 3 
1 4 6 
7 5 8 

1 2 3 
  4 6 
7 5 8 

1 2 3 
4   6 
7 5 8 

1 2 3 
4 5 6 
7   8 

1 2 3 
4 5 6 
7 8   

Movimientos: 18
Estados explorados: 191
Profundidad de la solución: 18
Profundidad máxima: 18
Tamaño lista de nodos frontera: 112
CPU times: user 26.4 ms, sys: 1.09 ms, total: 27.4 ms
Wall time: 25.6 ms
Caso 2:
Camino óptimo:
2 3 7 
1 4 8 
5 6   

2 3 7 
1 4 8 
5   6 

2 3 7 
1   8 
5 4 6 

2 3 7 
1 8   
5 4 6 

2 3   
1 8 7 
5 4 6 

2   3 
1 8 7 
5 4 6 

  2 3 
1 8 7 
5 4 6 

1 2 3 
  8 7 
5 4 6 

1 2 3 
5 8 7 
  4 6 

1 2 3 
5 8 7 
4   6 

1 2 3 
5   7 
4 8 6 

1 2 3 
5 7   
4 8 6 

1 2 3 
5 7 6 
4 8   

1 2 3 
5 7 6 
4   8 

1 2 3 
5   6 
4 7 8 

1 2

4. Compare la complejidad teórica del algoritmo A* vs. los resultados obtenidos en las corridas. 

La complejidad teórica del algoritmo A* es la siguiente [3]:

Tomando en cuenta que:

*  **b:** representa los estados visitados.
*  **d:** profundidad de la solución.
*  **D:** profundidad máxima.

Complejidad en tiempo: O(b^d)

Complejidad en espacio: O(b^d)

Al evaluar los valores de los casos probados tenemos que, los resultados obtenidos son menores a la complejidad teórica del algoritmo, por ejemplo para el caso 1:

*  **Caso 1:** 
       Tiempo: 29.8 ms (tiempo de ejecución)
       Espacio: 191 + 112 = 303 (espacio almacenado por la lista de nodos fronteras y lista de nodos explorados)
       
Estos resultados son menores que lo que muestra la complejidad teórica en tiempo y en espacioa, en donde, sustituyendo los valores tendría que quedar de la siguiente manera: O(191^18). Lo cual es mucho mayor a los resultados obtenidos.


5. Implemente una solución al problema 8-puzzle utilizando el algoritmo Iterative-deepening A* (IDA*), considerando la representación y heurística definidas anteriormente.

Al igual que en la implementación de A*, incorpore a su solución un conteo del número de estados visitados durante la búsqueda, el promedio de anchura/hijos del árbol de búsqueda, el tamaño de la solución obtenida y el tiempo de ejecución (para esto último utilice el *comando mágico* %%time de Jupyter).

Es importante destacar que en el código y toda implementación manejo el "blanco" en el 8-puzzle como el número 9. 

Además, otro punto importante es que antes de llamar o ejecutar las funciones necesarias para obtener la solución a través del algoritmo de IDA*, primero se ejecuta una función llamada **validateMatrix()**, la cual permite verificar si el tablero ingresado es un tablero válido o si por el contrario no tiene solución. En caso de que sea válido y tenga solución, ejecuta el algoritmo para resolver el problema.

Las ejecuciones del programa se realiza llamando a la función **mainIDA()** y se le pasa como parametro la matriz con el tablero del caso de prueba. Por ejemplo: mainIDA([[2,3,8],[7,4,1],[5,6,9]])

In [0]:
def B_IDA_star(node,limit):
     global repeat

     repeat = repeat + 1
     
     if node.h > limit:
          return node
     if goal(node.value):
          return node

     minimo = Node(None,[],10000,0,None,[])
     node.child = createNodes(node)

     for i in node.child:
          result = B_IDA_star(i,limit)

          if goal(result.value):
               return result
          if result.h < minimo.h:
               minimo.h = result.h

     return minimo
  
def IDA_star(matrix,pos):
     node = Node(matrix,pos,0,0,None,[])
     node.h = heuristic(node.value)
     limit = node.h

     while True:
          result = B_IDA_star(node,limit)

          if goal(result.value):
               return result
          if result.h < limit:
               return None

          limit = result.h


In [70]:
def mainIDA(matrix):

     global n
     global repeat
     global depth_max
      
     repeat = 0
     n = 3
     depth_max = 0
     depth_sol = 0

     pos = []

     for i in range(n):
          for j in range(n):
               if matrix[i][j] == 9:
                    pos = [i,j]
                    
     if validateMatrix(matrix,pos):
          node = IDA_star(matrix,pos)

          if(node != None):
               result = []
              
               depth_sol = node.level

               while node != None:
                    result.append(node)
                    node = node.parent

               result.reverse()

               print("Camino óptimo:")
               for i in result:
                    showMatrix(i.value)

               print("Movimientos:",len(result) - 1)
               print("Estados explorados:",repeat)
               print("Profundidad de la solución:",depth_sol)
               print("Profundidad máxima:",depth_max)
          else:
               print("No se encontró solución")
     else:
          print("Tablero invalido")
        
#CASOS DE PRUEBA:
print("Caso 1:")
%time mainIDA([[2,3,8],[7,4,1],[5,6,9]])
print("Caso 2:")
%time mainIDA([[2,3,7],[1,4,8],[5,6,9]])
print("Caso 3:")
%time mainIDA([[9,7,6],[1,2,8],[5,3,4]])
print("Caso 4:")
%time mainIDA([[3,8,6],[9,2,4],[7,1,5]])
print("Caso 5:")
%time mainIDA([[1,9,3],[5,2,6],[4,7,8]])
print("Caso 6:")
%time mainIDA([[7,2,4],[5,9,6],[8,3,1]])
print("Caso 7:")
%time mainIDA([[1,2,3],[8,9,4],[7,6,5]])


Caso 1:
Camino óptimo:
2 3 8 
7 4 1 
5 6   

2 3 8 
7 4   
5 6 1 

2 3   
7 4 8 
5 6 1 

2   3 
7 4 8 
5 6 1 

2 4 3 
7   8 
5 6 1 

2 4 3 
7 6 8 
5   1 

2 4 3 
7 6 8 
5 1   

2 4 3 
7 6   
5 1 8 

2 4 3 
7   6 
5 1 8 

2 4 3 
7 1 6 
5   8 

2 4 3 
7 1 6 
  5 8 

2 4 3 
  1 6 
7 5 8 

2 4 3 
1   6 
7 5 8 

2   3 
1 4 6 
7 5 8 

  2 3 
1 4 6 
7 5 8 

1 2 3 
  4 6 
7 5 8 

1 2 3 
4   6 
7 5 8 

1 2 3 
4 5 6 
7   8 

1 2 3 
4 5 6 
7 8   

Movimientos: 18
Estados explorados: 166
Profundidad de la solución: 18
Profundidad máxima: 18
CPU times: user 6.76 ms, sys: 3.13 ms, total: 9.88 ms
Wall time: 8.38 ms
Caso 2:
Camino óptimo:
2 3 7 
1 4 8 
5 6   

2 3 7 
1 4 8 
5   6 

2 3 7 
1   8 
5 4 6 

2 3 7 
1 8   
5 4 6 

2 3   
1 8 7 
5 4 6 

2   3 
1 8 7 
5 4 6 

  2 3 
1 8 7 
5 4 6 

1 2 3 
  8 7 
5 4 6 

1 2 3 
5 8 7 
  4 6 

1 2 3 
5 8 7 
4   6 

1 2 3 
5   7 
4 8 6 

1 2 3 
5 7   
4 8 6 

1 2 3 
5 7 6 
4 8   

1 2 3 
5 7 6 
4   8 

1 2 3 
5   6 
4 7 8 

1 2 3 
  5 6 
4 7 8 

1 2 3 
4 5 6 
  7

6. Compare los resultados obtenidos en las corridas de IDA* con los obtenidos en la ejecución de A. ¿Qué puede concluir?

Basándonos en los resultados obtenidos en las corridas del algoritmo A* y el IDA* podemos concluir que el IDA* se ejecuta más rápido que el A*, sin embargo, el A* visita muchos menos estados. Esto es debido a que el A* almacena estados en 2 listas, la lista de estados explorados es la lista de estados visitados y la lista de estados fronteras que son todos los posibles siguientes nodos de cada estado, y antes de visitar un nuevo estado verifica primero si éste se encuentra dentro de las listas mencionadas (explorados y fronteras) y en caso de que determine que ya ha visitado dicho estado anteriormente no lo visita, por lo cual no repite estados; en cambio, IDA* no almacena los estados visitados en alguna lista o algo por el estilo, por lo cual repite o visita de manera repetida muchos estados.

Además, otra diferencia importante es que, el A* al almacenar estados en listas, genera mucho más costo en memoria que el IDA* ya que el IDA* no almacena estados en memoria. 

Referencias:

[1] E. Scalise, "Técnicas de
búsqueda heurística", Centro de Ingeniería de Software y Sistemas (ISYS), Universidad Central de Venezuela, Venezuela, 2019

[2] Wikipedia, "Heurística admisible", https://es.wikipedia.org/wiki/Heur%C3%ADstica_admisible

[3] Wikipedia, "A* search algorithm", https://en.wikipedia.org/wiki/A*_search_algorithm


**Consideraciones:**  
- Agregue una celda de referencias y cite **todo** recurso que utilice. Utilice formato IEEE para las referencias.
- Esta tarea es individual (las restantes serán en grupos de dos personas).
- Para la entrega, hospede su solución en http://colab.research.google.com. La fecha y hora de la última modificación del documento será considerada el *timestamp* de entrega.
- Todas las celdas de código de la entrega deben estar ejecutadas para facilitar la corrección; sin embargo, se utilizará un set de casos de prueba para la revisión.
- La fecha de entrega de esta tarea es el lunes 24/06/2019, 11:59:59 PM (VET time). Debe enviar el URL por correo o mediante la opción de compartir del colab a escalise@gmail.com


**Posibilidad de Solución**  
Algunos estados iniciales del 8-puzzle no poseen solución. Existen varios métodos para verificar esto, uno de ellos se ilustra mediante un ejemplo.

Supongamos el siguiente estado inicial del 8-puzzle:

|&nbsp;|&nbsp;|&nbsp;|
|------|------|------|
| 1 | &nbsp; | 3 |
| 5 | 2 | 6 |
| 4 | 7 | 8 |

Para cada casilla con un valor _n&#8203;_, contamos cuántas casillas con valores menores a _n&#8203;_ se encuentran después de ésta (asumiendo _n&#8203;_=9 para la casilla vacía). Para el ejmplo mostrado arriba se tiene:

- Casilla con 1: no hay valores menores a 1 en ninguna casilla posterior
- Casilla con vacío (9): 7 valores menores a 9 (3, 5, 2, 6, 4, 7 y 8)
- Casilla con 3: un valor menor a 3 en casillas posteriores (2)
- Casilla con 5: 2 valores menores a 5 en casillas posteriores (2 y 4)
- Casilla con 2: no hay valores menores que lo sucedan.
- Casilla con 6: un valore menor a 6 en casillas posteriores (4)
- Casillas con 4, 7 y 8: no poseen valores menores posteriores.

Sumando los valores obtenidos se tiene: M = 0 + 7 + 1 + 2 + 0 + 1 + 0 + 0 + 0 = 11

Luego se calcula la distancia Manhatan entre la casilla vacía (1,2) y la posición que debe ocupar en el estado final (3,3). Esto es: D = |3 - 1| + |3 - 2| = 3.

Finalmente si la suma A + D es par, el 8-puzzle tiene solución; en caso contrario, no es resoluble. Para el ejemplo, A + D = 14, por lo tanto el 8-puzzle tiene solución.
